tycho-client/src/feed/mod.rs

```diff
@@ -212,34 +212,51 @@ impl SynchronizerStream {
 
     /// Tries to catch up a delayed state synchronizer.
     ///
-    /// If a synchronizer is delayed, this method will try to remove any as many waiting values in
-    /// it's queue until it caught up to the latest block, then it will try to wait for the next
-    /// expected block within a timeout. Finally the state is updated based on the outcome.
+    /// If a synchronizer is delayed, this method will try to catch up to the next expected block
+    /// by consuming all waiting messages in it's queue and waiting for any new block messages within
+    /// a timeout. Finally, all update messages are merged into one and returned.
     async fn try_catch_up(
         &mut self,
         block_history: &BlockHistory,
         max_wait: std::time::Duration,
     ) -> Option<StateSyncMessage> {
         let mut results = Vec::new();
         let extractor_id = &self.extractor_id;
-        while let Ok(msg) = self.rx.try_recv() {
-            let block_pos = block_history
-                .determine_block_position(&msg.header)
-                .unwrap();
-            results.push(Ok(Some(msg)));
-            if matches!(block_pos, BlockPosition::Latest) {
-                debug!(?extractor_id, "Extractor managed to catch up to latest state!");
-                break;
+
+        // Set a deadline for the overall catch-up operation
+        let deadline = std::time::Instant::now() + max_wait;
+
+        while std::time::Instant::now() < deadline {
+            match timeout(
+                deadline.saturating_duration_since(std::time::Instant::now()),
+                self.rx.recv(),
+            )
+            .await
+            {
+                Ok(Some(msg)) => {
+                    debug!(%extractor_id, block_num=?msg.header.number, "Received new message during catch-up");
+                    let block_pos = block_history
+                        .determine_block_position(&msg.header)
+                        .unwrap();
+                    results.push(msg);
+                    if matches!(block_pos, BlockPosition::NextExpected) {
+                        break;
+                    }
```

Comment:
Shouldn't we error in this case? I mean something must be really wrong if we got here.